pytorch_lightning/metrics/functional/classification.py

```diff
@@ -998,12 +1000,26 @@ def iou(
         tensor(0.4914)
 
     """
+    num_classes = get_num_classes(pred=pred, target=target, num_classes=num_classes)
+
+    # Determine minimum class index we will be evaluating. If using the background, then this is 0; otherwise, if
+    # removing background, use 1.
+    min_class_idx = 1 if remove_bg else 0
+
     tps, fps, tns, fns, sups = stat_scores_multiple_classes(pred, target, num_classes)
-    if remove_bg:
-        tps = tps[1:]
-        fps = fps[1:]
-        fns = fns[1:]
-    denom = fps + fns + tps
-    denom[denom == 0] = torch.tensor(FLOAT16_EPSILON).type_as(denom)
-    iou = tps / denom
-    return reduce(iou, reduction=reduction)
+
+    scores = torch.zeros(num_classes - min_class_idx, device=pred.device, dtype=torch.float32)
+    for class_idx in range(min_class_idx, num_classes):
```

Comment:
This seems like a slow approach if I have many classes?

Reply:
Good point. Instead of recomputing the `.any()` presence for target and pred for each class in the loop, in https://github.com/PyTorchLightning/pytorch-lightning/pull/3098/commits/c971b6135ae88b7b8cd3ec46d8342c601fbcd1db I now just use the already-computed support, true positives, and false positives.

Given the operations within the for loop here should otherwise be very quick, I still kept the for loop for handling the not-present check and per-class scores. Not sure if there's an easy and clear way around the for loop, but open to suggestions if you think we should continue pushing here.

Reply:
Marking this conversation as resolved, but feel free to re-open for further discussion or changes.